### ISSUE TITLE
1.1.124 reporting a type error when there should be none

### ISSUE THREAD

**Issue by ChristopherZhong** [microsoft/pyright#1678]
**Describe the bug**
In pyright `1.1.124`, a type error is reported whereas in version `1.1.123`, no type errors are reported.

**To Reproduce**
Steps to reproduce the behavior.

- Create a minimal file `test.py` as shown below

  `test.py`
  
  ```python
  from tempfile import mkstemp
  from zipfile import ZipFile
  
  _, temp_path = mkstemp()
  ZipFile(temp_path)
  ```

- Create a clean virtual environment without any additional packages
- Run pyright on the `test.py` file

The following is the type error reported by pyright `1.1.124`.

```
(venv) ➜  pyright-error npx pyright@1.1.124 test.py
npx: installed 1 in 1.368s
No configuration file found.
stubPath ~/pyright-error/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
~/pyright-error/test.py
  ~/pyright-error/test.py:5:9 - error: Argument of type "AnyStr@mkstemp" cannot be assigned to parameter "file" of type "str | _PathLike[str] | IO[bytes]" in function "__init__"
    Type "AnyStr@mkstemp" cannot be assigned to type "str | _PathLike[str] | IO[bytes]"
      Type "AnyStr@mkstemp" cannot be assigned to type "str"
      Type "AnyStr@mkstemp" cannot be assigned to type "_PathLike[str]"
      Type "AnyStr@mkstemp" cannot be assigned to type "IO[bytes]"
      Type "str | bytes" cannot be assigned to type "str | _PathLike[str] | IO[bytes]"
        Type "bytes" cannot be assigned to type "str | _PathLike[str] | IO[bytes]"
          "bytes" is incompatible with "str"
          "__fspath__" is not present (reportGeneralTypeIssues)
1 error, 0 warnings, 0 infos 
Completed in 0.622sec
```

**Expected behavior**
No type errors should be reported.
The expected output should be similar to the following output from pyright `1.1.123`.

```
(venv) ➜  pyright-error npx pyright@1.1.123 test.py
npx: installed 1 in 1.777s
No configuration file found.
stubPath ~/pyright-error/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
0 errors, 0 warnings, 0 infos 
Completed in 0.607sec
```

**Additional context**
The following is additional information.

```
(venv) ➜  pyright-error python --version
Python 3.8.2
```

```
pip --version
pip 19.2.3 from ~/pyright-error/venv/lib/python3.8/site-packages/pip (python 3.8)
```

```
(venv) ➜  pyright-error pip list
Package    Version
---------- -------
pip        19.2.3 
setuptools 41.2.0 
WARNING: You are using pip version 19.2.3, however version 21.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

**Comment by erictraut** [comment]
Thanks for the bug report. This is an interesting one that requires some explanation.

The problem stems from the typeshed type definition of the `mkstemp` function, which looks like this:
```python
def mkstemp(
    suffix: Optional[AnyStr] = ..., prefix: Optional[AnyStr] = ..., dir: Optional[_DirT[AnyStr]] = ..., text: bool = ...
) -> Tuple[int, AnyStr]: ...
```

Note that the return type includes the type variable `AnyStr` which needs to be "solved" based on the types of the arguments passed to the function. In this case, however, all of the parameters have default values, so it's valid to pass no arguments. That leaves the type variable `AnyStr` with no valid solution.

Earlier versions of pyright contained a bug where the default value of `...` was treated as `Any`. This meant that it incorrectly assumed that `AnyStr` should have an `Any` value, and the return type was determined to be `Tuple[int, Any]`. That's clearly incorrect.

With that bug fixed, pyright now returns an answer of `Tuple[int, AnyStr@mkstemp]`. In other words, it leaves the type variable "unsolved". It's not clear whether that's the correct behavior because PEP 484 does not indicate what should happen in this case. There's no great answer here. Perhaps the best solution is to replace the TypeVar with an `Unknown` type, which is a variant of `Any` used within pyright to signify a type that is implicitly unknown.

What does mypy do in this case? It determines that `AnyStr` should be a `str`. That's actually the correct answer, but it appears that mypy is getting this answer due to a bug. `AnyStr` is a constrained TypeVar, and mypy is simply returning the first constrained type. If the constraints used for `AnyStr` were reversed, it would provide a different answer. I've filed [this bug](https://github.com/python/mypy/issues/10241) to highlight the issue in mypy.

The typeshed stub is arguably the ultimate source of this bug. It is not providing a proper overload definition for the `mkstemp` function to cover the case where no arguments are provided. I'll look into submitting a change to address that issue.

**Comment by erictraut** [comment]
Here's the PR for typeshed: https://github.com/python/typeshed/pull/5133/files

**Comment by erictraut** [comment]
My change was accepted into the typeshed repo, and I've pulled the latest typeshed changes into pyright. The next release will address this issue.

**Comment by erictraut** [comment]
This is included in pyright 1.1.125, which I just published. It will also be included in the next release of pylance.
